### PR TITLE
coercion vec[Dictionary, Utf8] to Dictionary for coalesce function

### LINKED ISSUE #9925: Regression: Coercion stopped working for coalesce on a dictionary column
### Describe the bug

`coalesce` on a dictionary column worked on 36.0.0 but does not in 37.0.0

### To Reproduce

It works in 3.6.0 
```sql
DataFusion CLI v36.0.0

❯ create table test as values (arrow_cast('foo', 'Dictionary(Int32, Utf8)')), (null);
0 rows in set. Query took 0.035 seconds.

❯ select  coalesce(column1, 'none_set') from test;
+-----------------------------------------+
| coalesce(test.column1,Utf8("none_set")) |
+-----------------------------------------+
| foo                                     |
| none_set                                |
+-----------------------------------------+
2 rows in set. Query took 0.018 seconds.
```

Fails in 37.0.0:

```sql
DataFusion CLI v37.0.0
❯ create table test as values (arrow_cast('foo', 'Dictionary(Int32, Utf8)')), (null);
0 row(s) fetched.
Elapsed 0.021 seconds.

❯ select  coalesce(column1, 'none_set') from test;
Error during planning: No function matches the given name and argument types 'coalesce(Dictionary(Int32, Utf8), Utf8)'. You might need to add explicit type casts.
	Candidate functions:
coalesce(CoercibleT, .., CoercibleT)
```

### Expected behavior

Should get the same answer as 36.0.0

### Additional context
Maybe related to https://github.com/apache/arrow-datafusion/pull/9459

Kudos to @btasker for reporting this internally ❤

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #9925 

## Rationale for this change
The signature of the coalesce function has changed to variadic_equal and we need to add a new rule to coercion [Dictionary, Utf8] to Dictionary

### CODE REVIEW

File: datafusion/expr/src/type_coercion/functions.rs [hunk 1]
```diff
@@ -430,7 +430,7 @@ fn coerced_from<'a>(
         {
             Some(type_into.clone())
         }
-
+        Dictionary(_, _) if matches!(type_from, Utf8) => Some(type_into.clone()),
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/functions.rs:433]: I think this currently says it is possible to coerce from Utf8 to any dictionary type (e.g. even `Dictionary(Int64, Int32)`)

I think the `dictionary_coercion` function does something similar here: https://github.com/apache/arrow-datafusion/blob/37b73751b19b82516443579c714b3b5986dac927/datafusion/expr/src/type_coercion/binary.rs#L635-L634

So maybe we need to verify we can coerce the value types, like this?

```suggestion
        Dictionary(_, value_type) if coerce_from(value_type, from_type) => Some(type_into.clone()),
```

File: datafusion/sqllogictest/test_files/scalar.slt [hunk 2]
```diff
@@ -1779,6 +1779,18 @@ SELECT COALESCE(NULL, 'test')
...
+statement ok
+create table test1 as values (arrow_cast('foo', 'Dictionary(Int32, Utf8)')), (null);
+
+
+query ?
+select  coalesce(column1, 'none_set') from test1;
+----
+foo
+none_set
+
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/scalar.slt:1792]: Can we add a few more tests here?

1. The alternate (cast utf8 to dictionary):

```sql
select coalesce('none_set', column1) from test1;
select coalesce(NULL, 'none_set', column1) from test1;
```

3. A test on non string dictionaries, like this:
```sql
select coalesce(null, 34, arrow_cast(123, 'Dictionary(Int32, Int8)'));
select coalesce(null, arrow_cast(123, 'Dictionary(Int32, Int8)'), 34);
```
3. Also clear the table (e.g. `drop table test1`)?

**Review by alamb — COMMENTED** [review]
Thank you for working on this @Lordworms 🙏 

I think we need a few more tests and to handle a few more cases, but it looks good so far

File: datafusion/sqllogictest/test_files/scalar.slt (hunk 2 above)
Comment (Lordworms) [reply datafusion/sqllogictest/test_files/scalar.slt:1792]: sure

**Review by alamb — APPROVED** [review]
Thank you @Lordworms  -- this looks very nice to me 🙏

**Pull request merged**